src/builder.rs

```diff
@@ -631,20 +630,74 @@ impl Builder {
             self.bundle_type,
             self.spends,
             self.outputs,
+            self.burn,
         )
     }
 }
 
+/// Partition a list of spends and recipients by note types.
+/// Method creates single dummy ZEC note if spends and recipients are both empty.
+#[allow(clippy::type_complexity)]
+fn partition_by_asset(
+    spends: &[SpendInfo],
+    outputs: &[OutputInfo],
+    rng: &mut impl RngCore,
+) -> HashMap<AssetBase, (Vec<(usize, SpendInfo)>, Vec<(usize, OutputInfo)>)> {
+    let mut hm = HashMap::new();
+
+    for (i, s) in spends.iter().enumerate() {
+        hm.entry(s.note.asset())
+            .or_insert((vec![], vec![]))
+            .0
+            .push((i, s.clone()));
+    }
+
+    for (i, o) in outputs.iter().enumerate() {
+        hm.entry(o.asset)
+            .or_insert((vec![], vec![]))
+            .1
+            .push((i, o.clone()));
+    }
+
+    if hm.is_empty() {
+        let dummy_spend = SpendInfo::dummy(AssetBase::native(), rng);
+        hm.insert(
+            dummy_spend.note.asset(),
+            (vec![(usize::MAX, dummy_spend)], vec![]),
```

Comment:
Dummy spends/outputs should not participate in BundleMetadata indexing. To manage this, we need a way to mark them as a dummy in the indexed vectors of spends/outputs, so that in further code of `bundle` functions the following assignments will be skipped:
```
bundle_meta.spend_indices[spend_idx] = action_idx;
bundle_meta.output_indices[out_idx] = action_idx;
```
In the previous version of the code for this PR I simply used usize::MAX instead of the indices for marking dummy spends/outputs, so the checks:
```
if spend_idx < num_requested_spends {
if out_idx < num_requested_outputs {
```
skipped such items.

I agree that it was confusing, so now, I use the Option type to mark dummy items (for dummy items the index is None, for real ones - Some(...))

Reply:
nice, lets add comment
```rust
if hm.is_empty() {
        let dummy_spend = pad_spend(None, AssetBase::native(), rng);
        // dummy_spend should not be included in the indexing and marked as None.
        hm.insert(
            dummy_spend.note.asset(),
            (vec![(None, dummy_spend)], vec![]),
        );
```

Reply:
Fixed (added the comment).